README.md
[![npm version](https://badge.fury.io/js/@hms-networks%2Fkolibri-js-client.svg)](https://badge.fury.io/js/@hms-networks%2Fkolibri-js-client)

# Kolibri-JS-Client

The Kolibri-JS-Client is an easy to use Kolibri Consumer API client. It allows to interact with a Kolibri Broker with a simple request / response model with additional event driven data functionalities. The client is written in TypeScript and can also be used with JavaScript.

## Supported Kolibri Versions

Currently supported Kolibri Consumer Protocol Versions:

- v3.2

## Features

The Kolibri-JS-Client abstracts most of the Kolibri Protocol internals from the application and provides some additional / convenient features.

- Promise-based Request/Response API for the Kolibri Consumer Protocol over WebSockets.
- Event Listener support for receiving events and live data from the Kolibri Broker.
- Client side data merging for the kolibri.subscribe and kolibri.write live data.
- Client side Kolibri QoS Level and transaction handling
- Auto reconnect on connection loss
- NodeJS and Browser support
- Support for TLS
- Support for Kolibri-RPC
- Support for HTTP/HTTPS Proxy
- Fully typed

## Install

NPM

```bash
npm install @hms-networks/kolibri-js-client
```

Yarn

```bash
yarn add @hms-networks/kolibri-js-client
```

### Node and Browser Support

The client currently supports Browser Single Page applications and NodeJS environments.

HTML Apps can import the minimized bundle from the location build/bundles/kolibri-js-client.umd.min.js

```html
<script src="./build/bundles/kolibri-js-client.umd.min.js"></script>
```

## Getting Started

To be able to call kolibri RPCs the client has to be created with some configuration first. With the client object created you can then call Kolibri Consumer RPCs to the configured Kolibri Broker.

```typescript
import { ClientConfig, KolibriClientFactory } from '@hms-networks/kolibri-js-client';

// create config
const config: ClientConfig = {
     host: 'ws://localhost:8080',
     project: 'testv21',
     path: '/'
};
const client = KolibriClientFactory.create(config);

await client.connect();
// RPC calls...
await client.disconnect();
```

### Proxy Settings

Setting a proxy is only supported on Node environment. Proxy settings are ignored on the Browser.

```typescript
import { ClientConfig, KolibriClientFactory } from '@hms-networks/kolibri-js-client';

const config: ClientConfig = {
     host: 'ws://localhost:8080',
     project: 'testv21',
     path: '/',
     proxy: {
         host: 'localhost',
         port: 8089
     }
};

const client = KolibriClientFactory.create(config);
```

### TLS Connection

Custom TLS Configuration is only available on Node environment. On Browsers the config is ignored.

Tls connection with default options can be established by just providing host url with the wss:// protocol. If you want to configure the connection you can provide the optional tls config parameter to the ClientFactory.create method. For more details about the TlsConfig see <https://nodejs.org/api/https.html#https_https_request_options_callback>

```typescript
import { ClientConfig, KolibriClientFactory } from '@hms-networks/kolibri-js-client';

const config: ClientConfig = {
     host: 'ws://localhost:8080',
     project: 'testv21',
     path: '/',
     tls: {
        rejectUnauthorized: false
    }
};

const client = KolibriClientFactory.create(config);
```

### Samples Projects

The /sample directory contains examples how to setup and use some common scenarios.

## Sending RPCs

The client uses the request / response model to communicate with the Kolibri Broker. To send a kolibri RPC request call the desired function with the correct params object.

```typescript
const params: LoginParams = {
      user: 'Username',
      password: 'Password',
      interval: 60,
      timeout: 5
};
const result = await client.login(params);
```

## Event Listeners

Some kolibri RPCs trigger the Kolibri Broker to send data to the client. For that purpose the client is able to register event listeners to handle incoming data from the broker.

```typescript
client.addOnWriteListener((nodes: any[]) => {
 console.log(nodes);
});

const params: SubscribeParams = [
     { path: '/device01/temperatures/room1' }
];
const result = await client.subscribe(params);
```

After the registration of the addOnWriteListener and the client.subscribe RPC the Kolibri Broker will send data to the client. The client will then call the registered function with the data as the parameter. See Kolibri Consumer API Specification for more details.

## Error Handling

All the errors returned from the Kolibri Broker are instances of subclasses to the `KolibriRequestError` class containing the reason and additional data (if available).

```typescript
const params: LoginParams = {
      user: 'Invalid Username',
      password: 'Password',
      interval: 60,
      timeout: 5
};
try{
    const result = await client.login(params);
}
catch(e){
    if(e instanceof AccessDeniedError){
        console.log(e.message);
    }
}
```


CHANGELOG.md
# CHANGELOG

## 2.0.4

- chore: update dependencies
- fix: align to new jest's useFakeTimers api
- fix: remove unused mock

## 2.0.3

- chore: update dependencies

## 2.0.2

- fix: wss connection error with undefined tls config parameter

## 2.0.1

New release to fix npm registry issue

## 2.0.0

- [BC] Switch Kolibri Consumer Protocol to v3.2.c.kolibri
- [BC] Changed thrown error object

## 1.1.0

- add basic http/https proxy support
- update dependencies

## 1.0.0

Initial public release of the @hms-networks/kolibri-js-client lib


sample/oauth/main.ts
/*
* Copyright 2021 HMS Industrial Networks AB
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http: //www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/


import { ClientConfig, KolibriClientFactory, LoginParams, UpdateTokenParams } from '@hms-networks/kolibri-js-client';
import { Issuer } from 'openid-client';

async function main() {
    const keycloakIssuer = await Issuer.discover(
        'http://localhost:8081/auth/realms/hmshub' // issuer url
    );

    const oauthClient = new keycloakIssuer.Client({
        client_id: 'kolibri-js-client', // client id configured for your project
        token_endpoint_auth_method: 'none' // to send only client_id in the header
    });

    const config: ClientConfig = {
        host: 'ws://localhost:8080',
        project: 'management',
        path: '/'
    };

    let tokenSet = await oauthClient.grant({
        grant_type: 'password',
        username: 'user:project',
        password: 'password'
    });

    const client = KolibriClientFactory.create(config);
    await client.connect();

    let accessToken = tokenSet.access_token;

    const loginParams: LoginParams = {
        token: accessToken,
        interval: 60,
        timeout: 5
    };

    const result = await client.login(loginParams);

    // refresh token periodically and sent new token to broker
    setInterval(async () => {
        console.log('Getting new token and sending it to the Broker...');
        const refreshToken = tokenSet.refresh_token;
        tokenSet = await oauthClient.refresh(refreshToken!);

        const updateTokenParams: UpdateTokenParams = {
            token: tokenSet.access_token!
        };
        await client.updateToken(updateTokenParams);
    }, 10 * 1000);

    console.log(result);
}

main();

